### PR TITLE
Fixes for client non-persistence and blocking client.available()

### LINKED ISSUE #257: [WiFiServer/WiFiClient] Connection doesn’t persist when client object goes out of scope
### Boards

AMB82-Mini

### External Hardware

Other than included JXF37 camera, none.

### Hardware Configuration

None

### Version

latest main (checkout manually)

### IDE Name

VSCode w/ Arduino extension

### Operating System

Windows 11

### Auto Flash Mode

Disable

### Erase All Flash Memory (16MB)

Disable

### Standard Lib

Arduino_STD_PRINTF

### Upload Speed

2000000

### Description

Copying bug report in [AmebaIOT forum](https://forum.amebaiot.com/t/wifiserver-wificlient-connection-doesnt-persist-when-client-object-goes-out-of-scope/3177) to make reference issue for PR.
___________

According to Arduino’s [docs for server.available()](https://www.arduino.cc/reference/en/libraries/wifi/server.available/):

The connection persists when the returned client object goes out of scope; you can close it by calling client.stop().

However, we observed that the current behavior of WiFiServer and WiFiClient as it has been implemented into the Arduino SDK for the AMB82-Mini seems to contradict this.

### Sketch

```cpp
#include <WiFi.h>

char ssid[] = "";    // your network SSID (name)
char pass[] = "";        // your network password
int status = WL_IDLE_STATUS;
WiFiServer server(80);
WiFiClient client;

void setup()
{
    Serial.begin(115200);
    while (status != WL_CONNECTED) {
        status = WiFi.begin(ssid, pass);
        delay(5000);
    }
   
    server.begin();
}

void loop()
{
    client = server.available();
    if (client) {
        Serial.println("new client connected");
        String currentLine = "";
        while (client.connected()) {
            if (client.available()) {
                char c = client.read();
                Serial.write(c);
                if (c == '\n') {
                    if (currentLine.length() == 0) {
                        //do something
                    } else {
                        currentLine = "";
                    }
                } else if (c != '\r') {
                    currentLine += c;
                }
            }
        }
        client.stop();
        Serial.println("client disconnected");
    } else {
        Serial.println("waiting for client connection");
        delay(1000);
    }
}
```

### Error/Debug Message

```plain
[INFO] A client connected to this server :
[PORT]: 51193
[IP]:192.168.2.36

new client connected
client disconnected

[INFO] A client connected to this server :
[PORT]: 51192
[IP]:192.168.2.36

new client connected
client disconnected

[INFO] A client connected to this server :
[PORT]: 51194
[IP]:192.168.2.36

new client connected
client disconnected

// ...continues indefinitely
```

### Reproduce remarks

_No response_

### I have checked online documentation, FAQ, GitHub Wiki and existing/closed issues.

- [X] I confirm I have checked online documentation, FAQ, GitHub Wiki and existing/closed issues.

### LINKED ISSUE #258: [WiFiServer/WiFiClient] client.available() exhibits blocking behavior
### Boards

AMB82-Mini

### External Hardware

Other than included JXF37 camera, none.

### Hardware Configuration

None

### Version

latest main (checkout manually)

### IDE Name

VSCode w/ Arduino extension

### Operating System

Windows 11

### Auto Flash Mode

Disable

### Erase All Flash Memory (16MB)

Disable

### Standard Lib

Arduino_STD_PRINTF

### Upload Speed

2000000

### Description

This issue was created in order to create a reference issue for a PR.
____________

When `client.stop()` is called and then a new client is created soon after, `client.connected() == true` due to still pending socket close.

In the very commonly used context
```
while (client.connected()) {
    if (client.available()) {
        // other stuff happens
    }
}
```
this causes `client.available()` to block, which it should not do. When writing asynchronous programs, this can cause major issues in timing-sensitive tasks.

### Sketch

Requires fix for issue #257 to be applied because of WiFiClient copy assignment.

```cpp
#include <WiFi.h>

char ssid[] = "";    // your network SSID (name)
char pass[] = "";        // your network password
int status = WL_IDLE_STATUS;
WiFiServer server(80);
WiFiClient client;

void setup()
{
    Serial.begin(115200);
    while (status != WL_CONNECTED) {
        status = WiFi.begin(ssid, pass);
        delay(5000);
    }
  
    server.setNonBlockingMode(); // server is set to NonBlockingMode();
    server.begin();
}

void loop()
{
    client = server.available();
    if (client) {
        Serial.println("new client connected");
        String currentLine = "";
        while (client.connected()) {
            if (client.available()) {
                char c = client.read();
                Serial.write(c);
                if (c == '\n') {
                    if (currentLine.length() == 0) {
                        //do something
                    } else {
                        currentLine = "";
                    }
                } else if (c != '\r') {
                    currentLine += c;
                }
            }
        }
        client.stop();
        Serial.println("client disconnected");
    } else {
        Serial.println("waiting for client connection");
        delay(1000);
    }
}
```

### Error/Debug Message

```plain
// ... repeats while no client in non-blocking mode

[ERROR] get_available Accept connection failed
new client connected
client disconnected

[ERROR] get_available Accept connection failed
new client connected
client disconnected

[ERROR] get_available Accept connection failed
new client connected
client disconnected

[INFO] A client connected to this server :
[PORT]: 55775
[IP]:192.168.2.36

new client connected
GET / HTTP/1.1
Host: 192.168.2.172
Connection: keep-alive
Cache-Control: max-age=0
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9,fil;q=0.8
Cookie: __guid=259319114.1308697201864418800.1719734660203.9114

client disconnected

[INFO] A client connected to this server :
[PORT]: 55776
[IP]:192.168.2.36

new client connected

// blocked here by client.available()
```

### Reproduce remarks

Since the client does nothing in the sketch above, the socket can sometimes be closed quickly. But I was able to reproduce this issue reliably on first try or with one or two browser refreshes.

### I have checked online documentation, FAQ, GitHub Wiki and existing/closed issues.

- [X] I confirm I have checked online documentation, FAQ, GitHub Wiki and existing/closed issues.

### PR DESCRIPTION
**Pull request by redelacruz**
## Description of Change

- Fix: Fixes the broken rule of 3/5/0 in `WiFiClient` by removing the unneeded destructor, which caused the socket to stop prematurely on copy assignment.
- Fix: Fixes the blocking behavior caused by `client.available()` by removing the unneeded `try_again` label and its corresponding `goto`. In its place, we stop the socket (already pending stop) instead.
- API Updates: `WiFiClient` no longer has a destructor. This is congruent with Arduino's design for this API, as you must now call `client.stop()` to close the connection (see [[1](https://github.com/arduino/ArduinoCore-mbed/blob/a77715fffba3f471859956bc65a026a9c9ac1cda/libraries/WiFi/src/WiFiClient.h#L28)], [[2](https://github.com/arduino/ArduinoCore-mbed/blob/a77715fffba3f471859956bc65a026a9c9ac1cda/libraries/SocketWrapper/src/AClient.h#L27)], [[3](https://www.arduino.cc/reference/en/libraries/wifi/server.available/)]).

## Tests and Environments 
- AMB82-mini
- ambpro2_arduino V4.7.0
- VSCode v1.91.1
- Arduino (VSCode extension) v0.6.0
- Windows 11

## Remark related links
Closes #257, closes #258

### CODE REVIEW

**Review by M-ichae-l — COMMENTED** [review]
Thanks for the feedback. 
Would you consider to use `void WiFiClient::setBlockingMode()` and `void WiFiClient::setNonBlockingMode()` for the update? Keep both situation. This class private value may help '_is_blocked'

**Comment by redelacruz** [comment]
Hi, @M-ichae-l

Apologies for the delayed reply. I wasn't available for the last couple of days, and I wanted to confirm the following with my equipment before I replied to your comment:

Upon testing with my AMB82-Mini board, I was able to confirm that the proposed changes do not affect the operation of `void WiFiClient::setBlockingMode()` and `void WiFiClient::setNonBlockingMode()`. Both modes are working correctly with the patch I propose.

File: Arduino_package/hardware/libraries/WiFi/src/WiFiClient.cpp [hunk 1]
```diff
@@ -95,14 +90,14 @@ int WiFiClient::available()
         return 0;
     }
     if (_sock >= 0) {
-try_again:
```
Comment (M-ichae-l) [inline Arduino_package/hardware/libraries/WiFi/src/WiFiClient.cpp:98]: I would like to suggest to use `_is_blocked` by if condition, to keep both the update and original

File: Arduino_package/hardware/libraries/WiFi/src/WiFiClient.cpp [hunk 2]
```diff
@@ -66,11 +66,6 @@ WiFiClient::WiFiClient(uint8_t sock, tProtMode portMode, tBlockingMode blockMode
     _is_blocked = blockMode;
 }
 
-WiFiClient::~WiFiClient()
-{
-    stop();
```
Comment (M-ichae-l) [inline Arduino_package/hardware/libraries/WiFi/src/WiFiClient.cpp:71]: Is there any reason that remove this `stop` function?

Comment (ambiot) [reply Arduino_package/hardware/libraries/WiFi/src/WiFiClient.cpp:71]: How about the following update 

```
if (_is_blocked) {
    stop();
}
```

Comment (M-ichae-l) [reply Arduino_package/hardware/libraries/WiFi/src/WiFiClient.cpp:71]: @redelacruz   I think the suggested fix is ok. How about you?

**Comment by M-ichae-l** [comment]
@redelacruz 
The lasted commit fails at CI test.

**Pull request merged**